Add support for setting content in place of the title in a TitleBar

jonlipsky opened the pull request:
## Pull request type

Please check the type of change your PR introduces:

- [ ] Update
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes

## What is the current behavior?

Currently there is no way to set content in a `TitleBar` that is left justified and/or use content other than the default `TextBlock` for the title content. 

Issue Number: [#775] [#1025]

## What is the new behavior?

This PR adds support for (optionally) specifying content to be displayed as a replacement for the `TextBlock` which displays the title.  The following changes have been made:

- Added a `LeftHeader` property that can be optionally set to specify the view to be presented in place of the default `TextBlock` that displays the title (when set).
- A `TextBlock` is initialized in the `TitleBar` constructor and assigned to the `LeftHeader` property so that setting the title is still the default behavior.
- Renamed `Header` to `RightHeader` to make the API more self-documenting as to the purpose of the property.
- Updated the sandbox window example to demonstrate this new functionality.

<img width="928" alt="image" src="https://github.com/lepoco/wpfui/assets/1466386/3453e1b5-2fab-4210-9425-936993a44ef3">

## Other information

- This change is a breaking change from the previous version since `Header` has been renamed to `RightHeader`.  One remedy would be to add a new property called `Header`, which would delegate to `RightHeader` and be marked as deprecated to give users time to update their code.

- One additional enhancement that may be nice to have would be the ability to vertically align the Minimize, Maximize and Close buttons of the window.  It may be more aesthetically pleasing to have them center aligned depending on the content in the RightHeader.  As an example, JetBrains Rider does this in their custom titlebar:

<img width="1153" alt="image" src="https://github.com/lepoco/wpfui/assets/1466386/a23bda85-e2e0-4bdf-a467-94e9b4a3f03f">

jonlipsky committed:
Add support for setting content in place of the title in a TitleBar

537mfb commented:
To prevent backward compatibility issues you could have kept Header and just have added LeftHeader.

Once there's a major version change that can break backwards compatibility then a name change from Header to RightHeader could be considered without issue

jonlipsky commented:
@537mfb Yes, I'd be happy to submit an update to the pull request to change the naming of `RightHeader` back to `Header`.  I renamed that property because the original behavior was not obvious (in my personal opinion).

chucker commented:
Given that `Header`'s behavior changed between 2.x and 3.x, and this would change it back, I would argue the rename makes sense. Giving it a new name makes it clear that the behavior has changed.

Jay-o-Way commented:
I think you guys are forgetting that "left" and "right" _should_ be swapped when the OS is using a right-to-left language. Better use something like "primary" and "secondary, or something like that.

Jay-o-Way commented:
> * the ability to vertically align the Minimize, Maximize and Close buttons

That would be setting the **PreferredHeightOption** to **TitleBarHeightOption.Tall**
https://learn.microsoft.com/windows/apps/develop/title-bar#tall-title-bar-support-for-custom-title-bars

Jay-o-Way commented on src/Wpf.Ui/Controls/TitleBar/TitleBar.xaml:
```diff
@@ -118,42 +118,41 @@
+                            <!--  Title text or other header content  -->
                             <ContentPresenter
                                 Grid.Column="0"
+                                HorizontalAlignment="Left"
```
No need for this property when width=auto

jonlipsky commented:
@Jay-o-Way I initially called the renamed properties `LeadingHeader` and `TrailingHeader`; however it wasn't clear to me that the titlebar class was adaptive for right-to-left languages the way it is built, thus I renamed them.   I agree that if it is adaptive, then they should have an alternate name.

537mfb commented:
Not sure how easy this would be to achieve but how about having Header on the side of the system buttons (as is now) and make Title work like Content in other controls where if you give a string it works like a textblock but you can give it other arbitrary content.

Think of a Label - you can give it a string or a layout with several controls as content

jonlipsky commented:
@537mfb That's essentially what my PR does right now.  By default, it's a `TextBlock`, however if you set `LeftHeader`, that content becomes whatever you set it to.

537mfb commented:
@jonlipsky my idea is slightly different.

you could have:

`<iu:TitleBar Title="My App Name" ...`

and it would work as it does now. But you could also do:

```
<iu:TitleBar ....>
    <iu:TitleBar.Title>
         <StackPanel Orientation="Horizontal">
             <ui:SymbolIcon Symbol="Fluent24"/>
             ....
         </StackPanel>
    </iu:TitleBar.Title>
</iu:TitleBar>
```

pomianowski commented:
To be honest, I'm not a fan of the division into HeaderLeft and HeaderRight. I would prefer one big Header as a whole Titlebar up to navigation buttons with default value - Icon and Application Name

Summpot commented:
Any update?

chucker commented:
> To be honest, I'm not a fan of the division into HeaderLeft and HeaderRight. I would prefer one big Header as a whole Titlebar up to navigation buttons with default value - Icon and Application Name, similar to @537mfb's suggestion

[MAUI apparently now uses](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-titlebar-control-and-windowtitlebar-for-windows) the notion of `TrailingContent`, which is roughly what `HeaderRight` is in this PR:


![](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/media/titlebar-overview.png)

jonlipsky commented:
If a project admin/leader lets me know their preference as to how they want it to work, I'd happily amend the PR to implement that.  I took the initial approach I did because there were other issues/requests asking for it that way.

SamKr commented:
> If a project admin/leader lets me know their preference as to how they want it to work, I'd happily amend the PR to implement that. I took the initial approach I did because there were other issues/requests asking for it that way.

@pomianowski is the project owner, so I suppose [this reply](https://github.com/lepoco/wpfui/pull/1030#issuecomment-2053958429) counts?

This PR is awesome though, thanks for creating it @jonlipsky.

pomianowski committed:
Merge branch 'main' into development

pomianowski committed:
Use Header instead of HeaderLeft and TrailingContent instead of HeaderRight

Refactor TitleBar properties for clarity and usability

Renamed properties in the `TitleBar` class from `HeaderLeft` and `HeaderRight` to `Header` and `TrailingContent`. Updated XAML and C# files to reflect these changes, streamlining the structure of the `TitleBar`. Modified the constructor to set the `_titleBlock` to the new `Header` property. Adjusted hit testing logic in `WM.NCHITTEST` to accommodate the new property names, ensuring proper mouse interaction handling. These changes improve the overall clarity and usability of the `TitleBar` component.
